### PR TITLE
Validating HTTPRouteFilter type consistency

### LINKED ISSUE #942: Validate HTTP filter discriminated union consistency.
`HTTPRouteFilter` is a discriminated union where the `Type` field is the discriminator. The admission controller should validate that the only the matching field is present.

### PR DESCRIPTION
**Pull request by crmejia**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds validates that a HTTPRouterFilter Type matches it's value.

**Which issue(s) this PR fixes**:
Fixes #942

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by crmejia** [comment]
Hey @jpeach, this PR looks to address  #942

File: Makefile
Comment (jpeach) [inline Makefile:45]: Bad merge?

Comment (crmejia) [reply Makefile:45]: whoops..

File: apis/v1alpha2/validation/httproute.go
Comment (tokers) [inline apis/v1alpha2/validation/httproute.go:176]: These if logics might not so feasible, if a new filter type is introduced, we have to change a few places, and it's easy to forget.

Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go:176]: I see what you mean. It's hard to mantain. I've researched alternatives:
1. I started by looking at ways to implement discriminated unions in Go and they usually point to using interfaces(see [Alternatives to sum types in Go](https://making.pusher.com/alternatives-to-sum-types-in-go/), [StackOverflow](https://stackoverflow.com/questions/21553398/best-practice-for-unions-in-go)). Correct me If I'm wrong but I don't think it's right to add interfaces to the API just to solve this problem. 
2. Something that occurred to me is to compare the json tags with the type but I think is  "hacky".

Do you have any ideas or tips for the correct implementation? I have a hunch that there is a more elegant way to solve this and I just didn't look in the right place.

Comment (tokers) [reply apis/v1alpha2/validation/httproute.go:176]: If we can relax the constraints, say unused filter config can still be reserved there, than the switch statement can be simple.

Comment (jpeach) [reply apis/v1alpha2/validation/httproute.go:176]: The best I could come up with was this:

```Go
func validateUnionField(discriminator string, name string, value interface{}) {
        var isNil bool

        switch value := value.(type) {
        case *gatewayv1a2.HTTPRequestHeaderFilter:   
                isNil = (value == nil)
        case *gatewayv1a2.HTTPRequestMirrorFilter:
                isNil = (value == nil)
        case *gatewayv1a2.HTTPRequestRedirectFilter:
                isNil = (value == nil)
        case *gatewayv1a2.LocalObjectReference:
                isNil = (value == nil)
        default:
                // Puke an error so someone updates this code ...
        
        }

        if discriminator == name {
                if isNil {               
                        // Filter spec is not set, but it should be.
                }
        } else {
                if !isNil {
                        // Filter spec is set, but it should not be.
                }
        }
}

validateUnionField(filter.Type, gatewayv1a2.HTTPRouteFilterExtensionRef, filter.ExtensionRef)
validateUnionField(filter.Type, gatewayv1a2.HTTPRouteFilterRequestHeaderModifier, filter.RequestHeaderModifier)
         
```

Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go]: Made the changes. Does this look good to you?

Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go]: I've relaxed the constraints to make the switch statement simpler. I think that `validateUnionField`  is more complex and still has the issue of referencing the types so new types need to be added.

File: apis/v1alpha2/validation/httproute.go
Comment (robscott) [inline apis/v1alpha2/validation/httproute.go:171]: It doesn't look like this is actually being called anywhere other than the test. I think the best place to fit this in would be in `validateHTTPRouteUniqueFilters`, with that func renamed to `validateHTTPRouteFilters`

File: apis/v1alpha2/validation/httproute.go
Comment (robscott) [inline apis/v1alpha2/validation/httproute.go:177]: I think it would be helpful to have more specific error messages for each of these.
```suggestion
			return append(errs, field.Invalid(path.Child("filters"), path, "extensionRef must be specified for ExtensionRef filters"))
```

File: apis/v1alpha2/validation/httproute_test.go
Comment (robscott) [inline apis/v1alpha2/validation/httproute_test.go:635]: I think the actual path you'd want here would be something like `spec.rules[0].filters[0]`, but since this is just a test, an empty value is probably sufficient:

```suggestion
			errs := validateHTTPRouteFilterTypeMatchesValues(tc.routeFilter, field.NewPath(""))
```

**Review by robscott — COMMENTED** [review]
Thanks for the work on this @crmejia!

File: apis/v1alpha2/validation/httproute.go
Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go:171]: Would `validateHTTPRouteSpec` be a better fit to add the test? I noticed that `validateHTTPRouteUniqueFilters` is being called there along with `validateHTTPRouteBackendServicePorts`.

Comment (robscott) [reply apis/v1alpha2/validation/httproute.go:171]: Yeah not sure where the best place is, something needs to get called in `validateHTTPRouteSpec`. My suggestion would be to make that a generic function that validated a list of filters, including both uniqueness and that required fields were non-empty. That would also lead to another area for improvement, we would no longer need separate `validateHTTPRouteUniqueFilters` and `validateHTTPBackendUniqueFilters` functions.

Comment (jpeach) [inline apis/v1alpha2/validation/httproute.go:171]: you need to call this from somewhere, I suggest from `validateHTTPRouteSpec

File: apis/v1alpha2/validation/httproute.go
Comment (jpeach) [inline apis/v1alpha2/validation/httproute.go:177]:
```suggestion
			errs = append(errs, field.Invalid(path.Child("filters"), path, "type of filter must match type of value"))
```

Follow this pattern in the function, so that returning errs makes more sense.

Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go]: Oh yes, good catch. Thanks.

File: apis/v1alpha2/validation/httproute.go
Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go]: I did some refactoring. Please let me know what you think.

File: apis/v1alpha2/validation/httproute.go
Comment (crmejia) [inline apis/v1alpha2/validation/httproute.go:155]: That current error is coming from this switch. 
```
 apis/v1alpha2/validation/httproute.go:155:3: missing cases in switch of type v1alpha2.HTTPRouteFilterType: HTTPRouteFilterURLRewrite (exhaustive)
		switch filter.Type { 
```
But I'm not getting that error when I run this verification locally. Also, that switch statement has a `default` clause so it doesn't make sense to get this error. Any ideas what could be the reason?

File: apis/v1alpha2/validation/httproute.go
Comment (robscott) [inline apis/v1alpha2/validation/httproute.go:52]: This is already a great simplification, but can you take it one step further and just have a generic `validateHTTPRouteFilters` func instead that covers both of these? Would save us from looping through the same list of filters 2x.

File: apis/v1alpha2/validation/httproute.go
Comment (robscott) [reply apis/v1alpha2/validation/httproute.go:155]: It looks like this is complaining about a missing case for the `URLRewrite` filter type. That may have been added while you were working on this PR, so would've been easy to miss.

File: apis/v1alpha2/validation/httproute_test.go
Comment (robscott) [inline apis/v1alpha2/validation/httproute_test.go:572]: I wonder if we need to also make sure that the value of these fields is non empty? Even a simple check like `filter.RequestRedirect == nil || (gatewayv1a2.HTTPRequestHeaderFilter{}) == *filter.RequestRedirect)` might be helpful to ensure completely empty structs were also invalid.

**Review by robscott — COMMENTED** [review]
Thanks for your work on this @crmejia!

File: apis/v1alpha2/validation/httproute.go
Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go:155]: Great catch by the linter. I guess the compiler evaluates the `default:` as correct but the linter is a bit more clever. 👍🏾

File: apis/v1alpha2/validation/httproute.go
Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go:52]: I kept them separate to have a clear boundary for each validation but I can see the inefficiency.  I'll simplify 👍🏾

File: apis/v1alpha2/validation/httproute_test.go
Comment (crmejia) [reply apis/v1alpha2/validation/httproute_test.go]: Done. I had to use `reflect.DeepEqual` in some cases.

File: apis/v1alpha2/validation/httproute.go
Comment (jpeach) [inline apis/v1alpha2/validation/httproute.go:154]:
```suggestion
		if filter.ExtensionRef == nil || gatewayv1a2.LocalObjectReference{} == *filter.ExtensionRef {
```

I think that this is the correct pattern to apply here. We don't need deep equal, because if the filter is empty, the top-level fields will all equal the top level fields of the empty struct.

**Comment by jpeach** [comment]
Once the deep equal issue is done, this LGTM, pending Rob's concurrence.

/lgtm
/hold

File: apis/v1alpha2/validation/httproute.go
Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go:154]: I've changed these and others. However, in the case of `HTTPRequestHeaderFilter` it cannot be compared as it contains slices. Here is the error : `struct containing []ref cannot be compared`. So what I did instead was to check that those slices are not nil, instead of using deep equal.

Comment (jpeach) [reply apis/v1alpha2/validation/httproute.go:153]: OK, that sounds ok :+1:

File: apis/v1alpha2/validation/httproute.go
Comment (jpeach) [inline apis/v1alpha2/validation/httproute.go:157]: Need to complete this comment :)

File: apis/v1alpha2/validation/httproute.go
Comment (jpeach) [inline apis/v1alpha2/validation/httproute.go:162]:
```suggestion
		if filter.RequestMirror == nil || gatewayv1a2.HTTPRequestMirrorFilter{} == *filter.RequestMirror {
```

and the other places in this function

File: apis/v1alpha2/validation/httproute.go
Comment (jpeach) [inline apis/v1alpha2/validation/httproute.go:158]:
```suggestion
		if filter.RequestHeaderModifier == nil || (filter.RequestHeaderModifier.Add == nil && filter.RequestHeaderModifier.Set == nil && filter.RequestHeaderModifier.Remove == nil) {
```
IIUC the filter is missing if it is nil, or if all of the slices are empty.

Comment (crmejia) [reply apis/v1alpha2/validation/httproute.go]: removed the switch statement.

**Comment by crmejia** [comment]
After some discussion with Rob regarding validation we decided to check for `nil` instead of empty and to check these two cases for each filter:
1. `if filter != nil  && filter.Type != filter return some error`
2. `if filter == nil  && filter.Type == filter return some other error`
Please let me know what y'all think.

File: apis/v1alpha2/validation/httproute.go
Comment (robscott) [inline apis/v1alpha2/validation/httproute.go:148]: Tiny nit:
```suggestion
// validateHTTPRouteFilterTypeMatchesValue validates that only the expected fields are
// set for the specified filter type.
```

File: apis/v1alpha2/validation/httproute.go
Comment (robscott) [inline apis/v1alpha2/validation/httproute.go:183]: This is actually already covered by CRD validation, no need to also have it here: https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/httproute_types.go#L497-L498

**Review by robscott — COMMENTED** [review]
Thanks @crmejia! Just a couple tiny nits but otherwise LGTM.

**Comment by robscott** [comment]
Thanks @crmejia! Will leave final LGTM for someone else, but this looks great.

/approve

**Review by jpeach — APPROVED** [review]
/lgtm

\o/

**Pull request merged**
